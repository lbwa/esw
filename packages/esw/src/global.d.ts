declare type TranspileModuleType = 'cjs' | 'esm' | 'iife'
declare type TranspileModeEnv = 'development' | 'production'
