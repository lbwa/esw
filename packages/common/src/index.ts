export * from './types'
export * from './load'
export * from './utils'
export * from './table'
export * from './serialize-size'
export * as stdout from './stdout'
export * from './esbuild-msg'
