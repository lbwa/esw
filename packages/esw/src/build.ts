import path from 'path'
import fs from 'fs'
import { build as esbuild, BuildOptions, BuildResult } from 'esbuild'
import { map, of, pipe, switchMap, tap } from 'rxjs'
import externalEsBuildPlugin from './plugins/external'

function normalizeBuildOptions(cwd: string) {
  return pipe(
    // resolve package.json from cwd
    map((raw: BuildOptions) => {
      const pkgJsonPath = path.resolve(cwd, './', 'package.json')
      if (!fs.existsSync(pkgJsonPath)) {
        throw new Error(`package.json is required in the ${pkgJsonPath}`)
      }
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const pkgJson = require(pkgJsonPath) as Record<string, unknown>
      const options: BuildOptions = {
        bundle: true,
        ...raw
      }
      return {
        options,
        pkgJson
      }
    }),
    // resolve outfile and outdir from package.json field
    tap(({ options, pkgJson }) => {
      const outPath = (pkgJson['main'] ?? pkgJson['module']) as string
      const outFile = path.basename(outPath).replace(path.extname(outPath), '')
      const outDir = path.basename(path.dirname(outFile))
      options.outfile ??= outFile
      options.outdir ??= outDir
    })
  )
}

function applyExternalPlugin() {
  return pipe(
    tap(
      ({
        options,
        pkgJson
      }: {
        options: BuildOptions
        pkgJson: Record<string, unknown>
      }) => {
        const peerDeps = pkgJson['peerDependencies'] as Record<string, string>
        const deps = pkgJson['dependencies'] as Record<string, string>
        const groups = ([] as string[]).concat(
          Object.keys(peerDeps),
          Object.keys(deps)
        )
        options.plugins = (options.plugins ?? []).concat(
          externalEsBuildPlugin(groups)
        )
      }
    )
  )
}

function invokeEsBuildBuild<V extends { options: BuildOptions }>() {
  return pipe(switchMap(({ options }: V) => esbuild(options)))
}

function build(
  options: BuildOptions = {},
  cwd: string = options.absWorkingDir || process.cwd()
) {
  const build$ = of(options).pipe(
    normalizeBuildOptions(cwd),
    applyExternalPlugin(),
    invokeEsBuildBuild()
  )
  return new Promise<BuildResult>((resolve, reject) => {
    build$.subscribe({
      next: resolve,
      error: reject
    })
  })
}

export default build
