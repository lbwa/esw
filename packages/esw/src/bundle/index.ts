export * from './service'
export { createInference } from './options'
