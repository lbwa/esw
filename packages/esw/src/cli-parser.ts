import path from 'path'
import arg from 'arg'
import {
  EMPTY,
  from,
  of,
  pipe,
  throwError,
  zip,
  tap,
  map,
  concatMap,
  catchError
} from 'rxjs'
import { printAndExit } from './shared/log'
import { ProcessCode as Code } from './shared/constants'

export type CommandRunner = (argv?: string[]) => void
type AvailableArgs = typeof availableArgs
type Commands = typeof COMMANDS

// eslint-disable-next-line
const pkgJson = require(path.resolve(__dirname, '../..', 'package.json'))
const DEFAULT_COMMAND_NAME = 'build'
const COMMANDS: Record<string, () => Promise<CommandRunner>> = {
  build: () => import('./cli/build').then(({ default: run }) => run)
}
const availableArgs = {
  '--version': Boolean,
  '--help': Boolean,

  // alias
  '-v': '--version',
  '-h': '--help'
}

function printVersion<V extends { args: arg.Result<AvailableArgs> }>(
  pkg: Record<string, unknown>
) {
  return pipe(
    tap(({ args }: V) => {
      if (args['--version']) {
        printAndExit(`v${pkg['version'] as string}`, Code.OK)
      }
    })
  )
}

function printHelpIntoTerminal(commands: Commands) {
  const names = Object.keys(commands)
  printAndExit(
    `
    Usage
      $ esw <command>

      Available commands
        ${names.join(', ')}

      Options
        --version, -v Show version number
        --help, -h Display help messages

      For more information run a command with the --help flag
        $ esw ${names[0] ?? ''} --help
  `,
    Code.OK
  )
}

function printHelp<
  V extends { isValidCommand: boolean; args: arg.Result<AvailableArgs> }
>(commands: Commands) {
  return pipe(
    tap(({ isValidCommand, args: { '--help': help } }: V) => {
      if (!isValidCommand && help) {
        printHelpIntoTerminal(commands)
      }
    })
  )
}

function setEnv<V extends { commandName: string }>() {
  return pipe(
    tap(({ commandName }: V) => {
      const normalizedEnv =
        commandName === 'build' ? 'production' : 'development'
      process.env['NODE_ENV'] ||= normalizedEnv

      process.on('SIGTERM', () => process.exit(0))
      process.on('SIGINT', () => process.exit(0))
    })
  )
}

function normalizeArgs<
  V extends { isValidCommand: boolean; args: arg.Result<AvailableArgs> }
>(defaultCommandName: string) {
  return pipe(
    map(({ isValidCommand, args }: V) => {
      const commandName = isValidCommand ? args._[0] : defaultCommandName
      const forwardArgs = isValidCommand ? args._.slice(1) : args._
      if (args['--help']) {
        forwardArgs.push('--help')
      }
      return {
        forwardArgs,
        commandName
      }
    })
  )
}

function invokeCommand<
  V extends { commandName: string; forwardArgs: string[] }
>() {
  return pipe(
    concatMap(({ commandName, forwardArgs }: V) =>
      zip(from(COMMANDS[commandName]()), of(forwardArgs))
    ),
    map(([run, args]) => run(args))
  )
}

export default of(process.argv.slice(2)).pipe(
  map(argv =>
    arg(availableArgs, {
      //https://github.com/vercel/arg/blob/5.0.0/index.js#L13
      argv,
      permissive: true
    })
  ),
  catchError((err: Error & { code: string }) => {
    if (err.code === 'ARG_UNKNOWN_OPTION') {
      printHelpIntoTerminal(COMMANDS)
      return EMPTY
    }
    return throwError(() => err)
  }),
  map(args => ({ isValidCommand: !!COMMANDS[args._[0]], args })),
  printVersion(pkgJson),
  printHelp(COMMANDS),
  normalizeArgs(DEFAULT_COMMAND_NAME),
  setEnv(),
  invokeCommand()
)
