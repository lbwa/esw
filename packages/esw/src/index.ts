export { default as build } from './build/node'
export { default as watch } from './watch/node'
