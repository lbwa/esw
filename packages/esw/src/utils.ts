export function printAndExit(msg: string, code = 1) {
  // eslint-disable-next-line no-console
  console[code === 0 ? 'log' : 'error'](msg)
  process.exit(code)
}
