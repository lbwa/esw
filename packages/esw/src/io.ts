import { promises as fs } from 'fs'
import path from 'path'
import { BuildOptions } from 'esbuild'
import { isDef } from '@eswjs/common'

export async function findAllStaleDir(options: BuildOptions[]) {
  const dirs = [] as string[]
  await Promise.all(
    options.map(async ({ outdir, absWorkingDir = process.cwd() }) => {
      if (
        !outdir ||
        path.isAbsolute(outdir) ||
        path.resolve(absWorkingDir, outdir) === path.resolve(absWorkingDir)
      )
        return

      const stat = await fs.lstat(outdir).catch(() => null)
      if (isDef(stat) && stat.isDirectory() && !dirs.includes(outdir)) {
        dirs.push(outdir)
      }
    })
  )
  return dirs
}

export function rmDirs(dirs: string[]) {
  return Promise.all(
    dirs.map(dir => fs.rm(dir, { recursive: true, force: true }))
  )
}

export async function isFileExists(filepath: string) {
  try {
    const stat = await fs.stat(filepath)
    return stat.isFile()
  } catch (error) {
    return false
  }
}
