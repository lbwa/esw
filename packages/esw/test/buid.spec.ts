import path from 'path'
import fs from 'fs'
import { build } from '../src'

function getTestName() {
  return expect.getState().currentTestName.replace(/\s/g, '-')
}

function resolveFixture(name: string) {
  return path.resolve(__dirname, `./fixture/${name}`)
}

describe('node api - build', () => {
  it('should be works with package.json', async () => {
    const result = await build({
      absWorkingDir: resolveFixture('typescript'),
      logLevel: 'debug'
    })
    expect(result.errors).toHaveLength(0)
    expect(result.warnings).toHaveLength(0)
    expect(result).toMatchSnapshot(getTestName())

    expect(
      fs.existsSync(resolveFixture('typescript/dist/index.js'))
    ).toBeTruthy()
  })
})
