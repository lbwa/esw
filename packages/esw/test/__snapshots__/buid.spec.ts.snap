// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node api - build should be works with package.json: node-api---build-should-be-works-with-package.json 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;
